Add commit() method for multiple incremental updates without reloading

erano067 opened the pull request:
- Add commit() method to PDFDocument enabling chained incremental updates
- Fix font duplication: change 'modified' flag to 'alreadyEmbedded' in PDFFont
- Use PDFWriter instead of PDFStreamWriter for incremental saves
- Add comprehensive tests for commit() functionality (19 test cases)

## What?

This PR adds a `commit()` method to `PDFDocument` that enables multiple incremental updates to a PDF without reloading the document.

### The Problem

Currently, after calling `saveIncremental()`, you cannot perform another incremental save because:
1. The internal `originalBytes` still points to the initial PDF
2. The `prevStartXRef` hasn't been updated to the new XREF position
3. The snapshot used for change detection is stale

### The Solution

The new `commit()` method saves an incremental update and refreshes internal state, returning the complete PDF:

```typescript
// Load a PDF for incremental updates
const pdfDoc = await PDFDocument.load(pdfBytes, { forIncrementalUpdate: true });

// First update
const page = pdfDoc.getPage(0);
page.drawText('First update');
const firstCommit = await pdfDoc.commit();  // Returns complete PDF with update appended

// Second update - works without reloading!
page.drawText('Second update', { y: 100 });
const secondCommit = await pdfDoc.commit();  // Returns PDF with both updates

// Each commit builds on the previous, creating a proper XREF chain
```

### API

```typescript
/**
 * Commit the current changes to the document as an incremental update.
 * This allows you to save multiple incremental updates without reloading the PDF.
 *
 * @param options - Optional save options (same as saveIncremental)
 * @returns Promise<Uint8Array> - The complete PDF bytes including all updates
 */
async commit(options?: IncrementalSaveOptions): Promise<Uint8Array>
```

### Additional Fixes

- **Font duplication fix**: Changed `modified` flag to `alreadyEmbedded` in `PDFFont` to prevent duplicate font objects when embedding fonts across multiple commits
- **Writer consistency**: `saveIncremental()` now always uses `PDFWriter` instead of `PDFStreamWriter` to ensure proper incremental update format

## Why?

Real-world applications need to perform multiple incremental updates:

1. **Collaborative editing**: Multiple users adding annotations in sequence
2. **Form workflows**: Progressive form filling across sessions
3. **Audit trails**: Each modification preserved as a separate layer
4. **Digital signatures**: Multiple signatures, each requiring its own incremental update

Without `commit()`, developers must reload the entire PDF after each `saveIncremental()`, which is inefficient and loses the in-memory document state (pages, fonts, images already loaded).

## How?

The `commit()` method performs these key operations:

1. **Saves incrementally**: Calls `saveIncremental()` to get the incremental bytes
2. **Combines bytes**: Concatenates `originalBytes` + `incrementalBytes` into a new `Uint8Array`
3. **Updates originalBytes**: Sets this combined array as the new baseline
4. **Updates pdfSize**: Adjusts context size for correct offset calculations
5. **Updates prevStartXRef**: Parses the new XREF position from the increment
6. **Takes new snapshot**: Captures current object state for future change detection
7. **Returns complete PDF**: Returns the combined bytes for saving/transmission

### Alternative Considered

Reloading the PDF via `PDFDocument.load()` after each save - rejected because:
- Loses all in-memory references (pages, fonts, embedded images)
- Requires re-embedding resources
- Much slower for large documents
- Poor developer experience

## Testing?

Added 17 comprehensive unit tests covering:

- ✅ Multiple sequential commits
- ✅ Image deduplication across commits
- ✅ Font deduplication across commits
- ✅ Page operations (add/remove) across commits
- ✅ Mixed operations in single commits
- ✅ XREF chain integrity verification
- ✅ Metadata preservation
- ✅ Object stream handling
- ✅ commit() after full save() resets state
- ✅ saveIncremental() after commit() works correctly
- ✅ Error handling for empty increments
- ✅ Signature field widget preservation
- ✅ Byte-for-byte preservation for signature validity

All 668 tests pass:
```
Test Suites: 65 passed, 65 total
Tests:       668 passed, 668 total
```

## New Dependencies?

No.

## Screenshots

N/A - This PR adds programmatic functionality, not visual changes.

## Suggested Reading?

Yes. Relevant PDF specification sections:
- **7.5.6 Incremental Updates** - Core specification for appending changes
- **7.5.4 Cross-Reference Streams** - XREF stream format
- **7.5.5 File Trailer** - `Prev` entry linking XREF chain

## Anything Else?

This builds upon the existing incremental update infrastructure in the cantoo-scribe fork. The implementation is minimal and focused - it only adds what's necessary to enable chained incremental saves.

### Breaking Changes

None. This is purely additive - existing code continues to work unchanged.

### Migration

No migration needed. Existing `saveIncremental()` usage remains valid. The new `commit()` method is opt-in for workflows requiring multiple incremental updates.

## Checklist
- [x] I read [CONTRIBUTING.md](https://github.com/Hopding/pdf-lib/blob/master/docs/CONTRIBUTING.md).
- [x] I read [MAINTAINERSHIP.md#pull-requests](https://github.com/Hopding/pdf-lib/blob/master/docs/MAINTAINERSHIP.md#pull-requests).
- [x] I added/updated unit tests for my changes.
- [x] I added/updated integration tests for my changes.
- [x] I [ran the integration tests](https://github.com/Hopding/pdf-lib/blob/master/docs/CONTRIBUTING.md#running-the-integration-tests).
- [x] I tested my changes in Node, Deno, and the browser.
- [x] I viewed documents produced with my changes in Adobe Acrobat, Foxit Reader, Firefox, and Chrome.
- [x] I added/updated doc comments for any new/modified public APIs.
- [x] My changes work for both **new** and **existing** PDF files.
- [x] I [ran the linter](https://github.com/Hopding/pdf-lib/blob/master/docs/CONTRIBUTING.md#running-the-linter) on my changes.

adnsistemas committed:
Fix in object streams usage on XREF generation, for incremental update and incremental generation. Parameter adding for incremental update saving, to force useObjectStreams preference

erano067 committed:
Add commit() method for multiple incremental updates without reloading

- Add commit() method to PDFDocument enabling chained incremental updates
- Fix font duplication: change 'modified' flag to 'alreadyEmbedded' in PDFFont
- Use PDFWriter instead of PDFStreamWriter for incremental saves
- Add comprehensive tests for commit() functionality (17 test cases)

erano067 committed:
Add signature preservation tests and PR documentation

adnsistemas commented:
Hi, I like it. The change you made to only using XRefWriter instead of XRefStreamWriter is already fixed (detects PDF version, that was the real issue). Let me see how to properly merge this the next few days.
Thanks.

erano067 committed:
fix: track inline objects for incremental saves

Previously, inline objects (arrays/dicts embedded directly in other
objects, not registered as indirect objects) would not be tracked
for incremental saves. This caused changes to inline arrays like
MediaBox to be lost.

The fix enhances PDFContext.registerObjectChange() to:
1. First check if the object is directly registered (indirect object)
2. If not, search through all indirect objects to find which one
   contains the inline object, and mark that parent for save

Also fixed lint-staged config that had broken shell commands.

Added test: 'tracks inline array modifications for incremental saves'

All 673 tests pass.

erano067 commented:
Thanks! On top of the `commit()` function fixes, I also added:

- **Change tracking** for `PDFArray` mutations (`push`, `insert`, `remove`, `set`) and `PDFRawStream.updateContents()`
- **Inline object fix** – objects like MediaBox arrays embedded directly in dicts weren't being saved; now the containing indirect object is found and marked
- **5 new tests** covering these cases

All 673 tests pass.

adnsistemas committed:
Merge branch 'cantoo-scribe-master' into partialsave

adnsistemas committed:
Minor optimizations

adnsistemas committed:
Fix in file version

adnsistemas commented:
Ok, I need some help here..
I did some minor optimizations to this PR, and merge the changes in the "Original PR" into this one (the changes in the PR for cantoo-scribe/pdf-lib). Now I have an "updated version" of this PR, but I don't know how to merge those changes into this PR, for later merging the PR into the cantoo-scribe-incremental-update branch, which is the PR into cantoo-scribe.
I tested the this modified PR and pass all tests, and as you mention there is no changes that affects existing code.
Another thing is that I'm merging this changes into my fork of pdf-lib (@adnsistemas/pdf-lib) but I don't like the "commit" name, I'm changing it to "saveAndContinue". Let me know which one you prefer when merging this changes to the cantoo-scribe PR.
Thanks.

adnsistemas committed:
Updated documentation

adnsistemas commented:
Ok, I managed to push the changes to the PR, pushing to your branch. Check you are ok with the changes and, if so, I merge it into the cantoo-scribe PR.
Thanks.

adnsistemas reviewed (approved):
I did minor optimizations. I'm ok with the PR. Confirm if you are Ok with the changes I made.

erano067 commented:
How about `commitIncremental`?
I saw your changes, I think we are ready to ship and if there will be issues/improvements in the future we can address them in a dedicated PR.

can you tell whats the holdup of merging it to the official `@cantoo/pdf-lib`?

adnsistemas commented:
> can you tell whats the holdup of merging it to the official `@cantoo/pdf-lib`?

No idea.
